homeassistant/components/mazda/lock.py

```diff
@@ -0,0 +1,100 @@
+
+    def __init__(self, client, coordinator, index):
+        """Initialize the Mazda entity."""
+        super().__init__(client, coordinator, index)
+        self._assumed_lock_state = None
+        self._assumed_lock_state_timestamp = None
+
+    @property
+    def name(self):
+        """Return the name of the entity."""
+        vehicle_name = self.get_vehicle_name()
+        return f"{vehicle_name} Lock"
+
+    @property
+    def unique_id(self):
+        """Return a unique identifier for this entity."""
+        return self.vin
+
+    @property
+    def is_locked(self):
+        """Return true if lock is locked."""
+        api_last_updated_timestamp_str = self.coordinator.data[self.index]["status"][
+            "lastUpdatedTimestamp"
+        ]
+        api_last_updated_timestamp = datetime.datetime.strptime(
+            api_last_updated_timestamp_str, "%Y%m%d%H%M%S"
+        ).replace(tzinfo=datetime.timezone.utc)
+
+        now_timestamp = dt_util.utcnow()
+
+        # We need to do an optimistic update here because the API takes a few minutes to update after sending a lock/unlock request
+        # This will use self._assumed_lock_state as long as it is less than 5 minutes old and is more current than the API response
+        if (
+            self._assumed_lock_state is not None
+            and self._assumed_lock_state_timestamp is not None
+            and self._assumed_lock_state_timestamp > api_last_updated_timestamp
+            and (
+                (now_timestamp - self._assumed_lock_state_timestamp)
+                < datetime.timedelta(seconds=LOCK_ASSUMED_STATE_PERIOD)
+            )
+        ):
+            return self._assumed_lock_state == STATE_LOCKED
+
+        # We didn't use self._assumed_lock_state, so use the value from the most recent API response
+        door_lock_status = self.coordinator.data[self.index]["status"]["doorLocks"]
+
+        vehicle_is_unlocked = (
+            door_lock_status["driverDoorUnlocked"]
+            or door_lock_status["passengerDoorUnlocked"]
+            or door_lock_status["rearLeftDoorUnlocked"]
+            or door_lock_status["rearRightDoorUnlocked"]
+        )
+
+        return not vehicle_is_unlocked
+
+    async def async_lock(self, **kwargs):
+        """Lock the vehicle doors."""
+        # Optimistically assume the door is locked, because the API takes a long time to reflect the update
+        self._assumed_lock_state = STATE_LOCKED
+        self._assumed_lock_state_timestamp = dt_util.utcnow()
+
+        self.schedule_update_ha_state()
+
+        await self.client.lock_doors(self.vehicle_id)
+
+    async def async_unlock(self, **kwargs):
+        """Unlock the vehicle doors."""
+        # Optimistically assume the door is unlocked, because the API takes a long time to reflect the update
+        self._assumed_lock_state = STATE_UNLOCKED
+        self._assumed_lock_state_timestamp = dt_util.utcnow()
+
+        self.schedule_update_ha_state()
```

Comment:
```suggestion
        self.async_write_ha_state()
```

From your comment it doesn't seem like your intent was to schedule an update, only write the state of the entity.

Reply:
Will `async_write_ha_state()` call the `is_locked()` function again in order to recalculate the current state? I need to make sure it recalculates the state so that it will use the `_assumed_lock_state` instead of the state from the API response.

Reply:
After some testing, it looks like it does call it again, so I went ahead and made this change. 👍